### PR TITLE
14.0 selection ondelete fix adt

### PR DESCRIPTION
**Pull request by Elkasitu**
Read commits for a description

### CODE REVIEW

File: odoo/addons/base/models/ir_model.py [hunk 1]
```diff
@@ -2149,13 +2171,11 @@ def delete(records):
...
 
-        # Remove fields, selections and relations. Note that the selections of
-        # removed fields do not require any "data fix", as their corresponding
-        # column no longer exists. We can therefore completely ignore them. That
-        # is why selections are removed after fields: most selections are
-        # deleted on cascade by their corresponding field.
-        delete(self.env['ir.model.fields'].browse(unique(field_ids)))
+        # ir.model.fields.selection must be deleted first and before any fields, SQL CASCADE
+        # deletes must be avoided if the ondelete actions are to be executed on the records
+        # pointing to the to-be-deleted ir.model.fields.selection
```
Comment (rco-odoo) [inline odoo/addons/base/models/ir_model.py:2176]: This is very obscure.  You could explain with an example.
Let me give it a try, to see whether I understood the issue.
> If we delete a selection field, and some of its values have `ondelete='cascade'`, we expect the records with that value to be deleted.  If we delete the field first, the column is dropped and the selection is gone, and thus the records above will not be deleted.

File: odoo/addons/test_new_api/models/test_new_api.py [hunk 2]
```diff
@@ -1152,6 +1152,29 @@ class SelectionNonStored(models.Model):
...
+    _inherit = 'test_new_api.model_selection_required_for_write_override'
+
+    my_selection = fields.Selection(selection_add=[
+        ('divinity', "Divinity: Original Sin 2"),
+    ], ondelete={'divinity': 'set default'})
+
+    def write(self, vals):
+        if 'my_selection' in vals:
+            raise ValueError("No... no no no")
+        super().write(vals)
```
Comment (rco-odoo) [inline odoo/addons/test_new_api/models/test_new_api.py:1175]: Thou shall `return`.

File: odoo/addons/base/models/ir_model.py [hunk 3]
```diff
@@ -1389,6 +1389,12 @@ def _process_ondelete(self):
                 selection._get_records().write({field.name: value})
             elif ondelete == 'cascade':
                 selection._get_records().unlink()
+            elif ondelete is None:
+                # special case for custom fields
+                if field.manual and not field.required:
+                    write(selection._get_records(), field.name, False)
+                # nothing to do, the selection does not come from a field extension
+                continue
```
Comment (rco-odoo) [inline odoo/addons/base/models/ir_model.py:1419]: Wouldn't it be clearer to let `ondelete` default to `'set null'` above?
```python
ondelete = (field.ondelete or {}).get(selection.value)
# special case for custom fields
if ondelete is None and field.manual and not field.required:
    ondelete = 'set null'

if ondelete is None:
    # nothing to do, the selection does not come from a field extension
    continue
elif callable(ondelete):
    ...
```

**Review by rco-odoo — COMMENTED** [review]
Commits are upside-down, which makes the first commit incorrect (function `write` is not defined yet).

File: odoo/addons/base/models/ir_model.py (hunk 3 above)
Comment (rco-odoo) [reply odoo/addons/base/models/ir_model.py:1419]: BTW, if you use my suggestion, you don't need to swap commits.

Comment (Elkasitu) [reply odoo/addons/base/models/ir_model.py]: Good catch, the special case for custom fields I discovered after pushing and creating the PR, so I simply added it in there without thinking much about it, this is definitely clearer.

Also originally it was a single commit but I thought splitting them would be best, but I missed the call to `write` indeed

File: odoo/addons/test_new_api/models/test_new_api.py (hunk 2 above)
Comment (Elkasitu) [reply odoo/addons/test_new_api/models/test_new_api.py]: Forgive me father, for I hath not `return`ed

File: odoo/addons/base/models/ir_model.py [hunk 4]
```diff
@@ -1368,6 +1368,28 @@ def unlink(self):
 
     def _process_ondelete(self):
         """ Process the 'ondelete' of the given selection values. """
+        def write(records, col, value):
+            if not records:
+                return
+            try:
+                records.write({col: value})
```
[the remark this answers is not in the document]
Comment (Elkasitu) [reply odoo/addons/base/models/ir_model.py]: I did it, but I'm not sure if it's necessary, I will double check but I think in the current 14.0 when the `raise` inside the `write` override is triggered, the cursor is not kept dirty

File: odoo/addons/base/models/ir_model.py [hunk 5]
```diff
@@ -1368,6 +1368,29 @@ def unlink(self):
...
+                    "Could not fulfill ondelete action for field %s of model %s, "
+                    "attempting ORM bypass...", col, records._name,
+                )
+                query = sql.SQL("UPDATE {} SET {}=%s WHERE id IN %s").format(
+                    sql.Identifier(records._table),
+                    sql.Identifier(col),
+                )
+                # if this fails then we're shit out of luck and there's nothing we can do
+                # except fix on a case-by-case basis
+                with self.env.cr.savepoint():
```
Comment (rco-odoo) [inline odoo/addons/base/models/ir_model.py:1390]: this savepoint is not useful here, as the caller does not expect the transaction to have rolled back

File: odoo/addons/base/models/ir_model.py [hunk 6]
```diff
@@ -1368,6 +1368,29 @@ def unlink(self):
...
+                    self.env.cr.execute(query, (value, tuple(records.ids)))
```
Comment (rco-odoo) [inline odoo/addons/base/models/ir_model.py:1391]: `value` should be converted to the SQL format (`False` → `None`)

**Review by rco-odoo — APPROVED** [review]
@robodoo rebase-ff r+

**Comment by rco-odoo** [comment]
@robodoo delegate+

**Comment by Elkasitu** [comment]
@robodoo rebase-ff r+

**Pull request closed without merging**